File: leetcode/78.子集.go
func subsets(nums []int) [][]int {
    //注意子集和全排列的区别，不必要非要等到数组遍历完毕才加入结果中
    ans := [][]int{}
    path := []int{}
    n := len(nums)
    var dfs func(int)
    dfs = func(start int){
        ans = append(ans, append([]int(nil), path...))
        for i := start; i < n; i++ {
            path = append(path, nums[i])
            dfs(i + 1)
            path = path[:len(path) - 1]
        }       
    }
    dfs(0)
    return ans
}

File: leetcode/78.子集.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        ans = []
        t = []
        # 输出答案的视角
        def dfs(inx):
            ans.append(t[::])
            if inx == n:
                return
            for j in range(inx, n):
                t.append(nums[j])
                dfs(j+1)
                t.pop()
        dfs(0)
        return ans